### PR TITLE
12 job splitters

### LINKED ISSUE #11: Implement post-job actions in GFMAP (Rasterization purpose)
Implement a way of triggering post-job actions in GFMAP. Once a openeo job is finished, its content is downloaded and the job task is performed locally based on the newly downloaded file.

@kvantricht This task is directly related to the generation of label rasters (rasterization process). For example, we can link this post-job with the extraction of sentinel-2 data

### PR DESCRIPTION
**Pull request by GriffinBabe**
Closes #11 and #12 

- Added job splitter using h3 library. Adding the h3 library and using it directly is more lightweight in terms of package and drastically more computationally efficient than having a pre-generated dataset containing the h3 grid. Especially when it comes to low resolutions.
- Added job manager, extending `MultiBackendJobManager`. The job manager also implements post-job actions that are computed on other threads to keep the status update smooth.

TODO before merging:

- [x] Polish and add a demo notebook to the branch

### CODE REVIEW

**Comment by soxofaan** [comment]
> @soxofaan I'm confused, there is a lint check that fails on black but the black check is disabled in the workflow file? 

That's indeed weird. What might be the case is Gihub runs the workflow from main (instead of your branch) because you are making a pull request targeting main.

**Comment by soxofaan** [comment]
> Do you think I should be using multiprocessing instead of multithreading? In case GFMAP is running on a VM with multiple CPUs that would really accelerate possibly heavy post-job actions

If I remember correctly, threading and multiprocessing have a common API, so I guess it's most future proof to stay within this common API, so that you can switch when one approach doesn't cut it

**Comment by GriffinBabe** [comment]
@kvantricht also corresponds to #43

**Comment by kvantricht** [comment]
> Problem remaining with the "to_scl_dilation_mask", except that everything should be ready for merge

is that something that needs to be fixed before merging or do we create another issue for that? What exactly was the problem?

File: src/openeo_gfmap/manager/job_splitters.py
Comment (kvantricht) [inline src/openeo_gfmap/manager/job_splitters.py:17]: Thought you mentioned `grid_resolution=3` to be the default? Need to sync this with @cbutsko. For WorldCereal it's likely that the splitting is already done in RDM, meaning that a sample already contains the gridbox ID of a specific resolution (currently 3) that allows easy querying of the RDM API to select only samples in one gridbox.
But for general functionality good to have it here too, we just might want to align the default resolution.

File: src/openeo_gfmap/stac/constants.py
Comment (kvantricht) [inline src/openeo_gfmap/stac/constants.py:6]: If this title is something we want a default value for, not sure if this has to be WorldCereal? At this point, functionality should be broader than just WorldCereal extraction use case right?

File: src/openeo_gfmap/stac/constants.py
Comment (kvantricht) [inline src/openeo_gfmap/stac/constants.py:7]: What does `L1` mean here?

File: src/openeo_gfmap/stac/constants.py
Comment (kvantricht) [inline src/openeo_gfmap/stac/constants.py:22]: is this defined at the level where all sensors are aggregated into one STAC collection? Do new sensors have to be added here? What's the typical workflow?

File: src/openeo_gfmap/stac/constants.py
Comment (kvantricht) [inline src/openeo_gfmap/stac/constants.py:92]: don't forget to update to new band names

**Review by kvantricht — COMMENTED** [review]
Huge work done, nice! I added some comments/questions. Don't entirely get the STAC yet and how much it's tuned now to WorldCereal. Also band names still need to be updated there. In the end is this just the example belonging to Sentinel-2 extraction for WorldCereal?

What I didn't check but what's important as you know is proper testing. I trust you've been writing the proper tests along the way.

File: src/openeo_gfmap/stac/constants.py
Comment (soxofaan) [reply src/openeo_gfmap/stac/constants.py:6]: indeed, we should WorldCereal or other concrete references in gfmap

Comment (GriffinBabe) [reply src/openeo_gfmap/stac/constants.py:6]: I removed them and made them as parameters for the GFMAPJobManager, I agree they shouldn't be hardcoded values

File: src/openeo_gfmap/stac/constants.py
Comment (GriffinBabe) [reply src/openeo_gfmap/stac/constants.py:7]: Removed, but that's something @VincentVerelst implemented

File: src/openeo_gfmap/stac/constants.py
Comment (GriffinBabe) [reply src/openeo_gfmap/stac/constants.py:104]: I renamed it to LABEL and will content myself with it. I think it's about time to merge this huge branch. Let's talk on monday about this because I'm not totally sure either @VincentVerelst  took care of that part

File: src/openeo_gfmap/stac/constants.py
Comment (GriffinBabe) [reply src/openeo_gfmap/stac/constants.py:89]: Done

File: src/openeo_gfmap/stac/constants.py
Comment (GriffinBabe) [reply src/openeo_gfmap/stac/constants.py:19]: To be discussed monday

**Review by kvantricht — APPROVED** [review]
I will approve for now. After Monday discussion we can merge this and add remaining issues if needed.

**Comment by GriffinBabe** [comment]
Created issue for to_scl_dilation_mask and for the STAC AssetDefinition

**Pull request merged**
